src/trace/persistent/cursor.rs

```diff
@@ -0,0 +1,284 @@
+    /// Current index of iterator in `cur_vals`.
+    ///
+    /// Value will be `0 <= val_idx < cur_vals.len()`.
+    val_idx: usize,
+
+    cursor_upper_bound: Option<B::Key>,
+    /// Temporary storage serializing seeked keys.
+    tmp_key: ReusableEncodeBuffer,
+}
+
+impl<'s, B> PersistentTraceCursor<'s, B>
+where
+    B: Batch + 's,
+    B::Key: Ord + Clone + Encode + Decode,
+    B::Val: Ord + Encode + Decode,
+    B::Time: Encode + Decode,
+    B::R: MonoidValue + Encode + Decode,
+{
+    /// Loads the current key and its values from RocksDB and stores them in the
+    /// [`PersistentTraceCursor`] struct.
+    ///
+    /// # Panics
+    /// - In case the `db_iter` is invalid.
+    fn update_current_key_weight(&mut self) {
+        assert!(self.db_iter.valid());
+        let (key, values) = PersistentTraceCursor::<'s, B>::read_key_val_weights(&mut self.db_iter);
+
+        self.val_idx = 0;
+        self.cur_key = key;
+        self.cur_vals = values;
+    }
+
+    /// Loads the current key and its values from RocksDB.
+    ///
+    /// # Returns
+    /// - The key and its values.
+    /// - `None` if the iterator is invalid.
+    #[allow(clippy::type_complexity)]
+    fn read_key_val_weights(
+        iter: &mut DBRawIterator<'s>,
+    ) -> (Option<B::Key>, Option<Values<B::Val, B::Time, B::R>>) {
+        loop {
+            if !iter.valid() {
+                return (None, None);
+            }
+
+            if let (Some(k), Some(v)) = (iter.key(), iter.value()) {
+                let (key, _) =
+                    decode_from_slice(k, BINCODE_CONFIG).expect("Can't decode current key");
+                let (values, _): (PersistedValue<B::Val, B::Time, B::R>, usize) =
+                    decode_from_slice(v, BINCODE_CONFIG).expect("Can't decode current value");
+                match values {
+                    PersistedValue::Values(vals) => {
+                        return (Some(key), Some(vals));
+                    }
+                    PersistedValue::Tombstone => {
+                        // Skip tombstones:
+                        // they will be deleted by the compaction filter at a later point in time
+                        iter.next();
+                        continue;
+                    }
+                }
+            } else {
+                // This holds according to the RocksDB C++ API docs
+                unreachable!("db_iter.valid() implies Some(key)")
+            }
+        }
+    }
+}
+
+impl<'s, B: Batch> PersistentTraceCursor<'s, B>
+where
+    B::Key: Clone + Ord + Encode + Decode,
+    B::Val: Ord + Encode + Decode,
+    B::Time: Encode + Decode,
+    B::R: Encode + Decode,
+{
+    /// Creates a new [`PersistentTraceCursor`], requires to pass a handle to
+    /// the column family of the trace.
+    pub(super) fn new(cf: &Arc<BoundColumnFamily>) -> Self {
+        let mut db_iter = ROCKS_DB_INSTANCE.raw_iterator_cf(cf);
+        db_iter.seek_to_first();
+        let (cur_key, cur_vals) =
```

Comment:
That's ok (the proptests check this) but I also added a unit test for it now (it will set `cur_key` and `cur_vals` to None https://github.com/vmware/database-stream-processor/blob/f075e32f67db1e156076c64f6550fa65fc9c9742/src/trace/persistent/cursor.rs#L78)